### DISCUSSION TITLE
[question] Building ksh93s 2006-12-22 and older

### DISCUSSION THREAD

**Discussion by JohnoKing** [ksh93/ksh#388] [Q&A] [answered]
This question pertains to the versions of ksh in the [ast-open-history](https://github.com/ksh93/ast-open-history) repo. While I was browsing through the versions there I noticed that most of the INIT build system is missing in all versions older than ksh93s+ 2007-03-01 (the [93s INIT source code](https://github.com/ksh93/ast-open-history/tree/2006-12-22/src/cmd/INIT) lacks the `package.sh` and `iffe.sh` source code, while [93s+](https://github.com/ksh93/ast-open-history/tree/2007-03-01/src/cmd/INIT) has both). Is this because of an accidental omission when the repo was made, or is it because you lack the INIT tarballs for these versions? As it is right now any attempt to build these versions requires obtaining INIT from another source. ([Fedora's archive](https://src.fedoraproject.org/lookaside/pkgs/rpms/ksh) and Slackware ([Ver. 8.0](https://mirrors.slackware.com/slackware/slackware-8.0/source/ap/ksh93/) + [Ver. 8.1](https://mirrors.slackware.com/slackware/slackware-8.1/source/ap/ksh93/) + [Ver 10.0](https://mirrors.slackware.com/slackware/slackware-10.0/source/ap/ksh93/)) have quite a few tarballs for INIT, but not all of them.)

**Comment by McDutchie** [comment, marked as answer]
Presumably that means it was missing from the original, as ast-open-history is simply the [archive of the ast-open repo that I've received from @ormaaj](https://drive.google.com/drive/folders/1KXRdIJ7_hdH8AEPEgY7Hyjkl4RY9ThTu). He sent it to me, I unpacked it, cd'ed into it, changed the upstream URL to ast-open-history at github, did a `git push`, and then told github to archive the repo. That's all I did and all I know about it.

That link comes with a README that explains how @ormaaj put it together. But it wouldn't surprise me if the ast-open repo was a mess to begin with. The AT&T folks made a mess of so many things.

I've also noticed the multishell repo has some ksh versions that ast-open-history doesn't -- the 2012-08-12 release included. But multishell doesn't include all the non-ksh AST stuff.

If someone cares enough to do it, it might be worth piecing together a new repo combining the various archival sources. For it to be really useful, that should be done in such a way that 'git blame' can instantly identify the version in which any line of code was introduced.

**Reply by JohnoKing** [reply]
Thanks for linking to the archive on Google Drive! I'll probably make a new ast-open repo for that sometime in the future, as it has many tarballs of ast-open 93v- the ast-open-history repo doesn't include (this will also make backporting fixes from 93v- much easier).

**Comment by ormaaj** [comment]
On Fri, Dec 17, 2021 at 10:45 AM Martijn Dekker
***@***.***> wrote:
>
> Presumably that means it was missing from the original, as ast-open-history is simply the archive of the ast-open repo that I've received from @ormaaj. He sent it to me, I unpacked it, cd'ed into it, changed the upstream URL to ast-open-history at github, did a git push, and then told github to archive the repo. That's all I did and all I know about it.
>
> That link comes with a README that explains how @ormaaj put it together. But it wouldn't surprise me if the ast-open repo was a mess to begin with. The AT&T folks made a mess of so many things.

The `ast-open-git.tar.bz2` file was only the original unmodified repo
that included historical snapshots up to a point, but is incomplete.
The `ast` subdirectory in my google drive archive however contains all
of the subsequent snapshots that I had collected from the att website,
mailing lists, and a few obscure code drops elsewhere as things were
falling apart.

If you take that repo and combine it with the big snapshot pile under
the `ast` dir then that should be essentially the complete history
(less perhaps a few i might be missing that you might dig up from the
multishell repo.

I believe they were trying to switch over to a git workflow internally
so they threw together that repo from a pile of historical snapshots
as a starting point, but that plan was abandoned and they just
continued uploading snapshot archives. You can see from the git log
that GSF did it all at once and didn't bother recreating the timeline
with correct release dates, so you have to look at the changelog
files.

I had been meaning to do that and wanted to do a nice job fixing up
the logs to include changelog content, but I wanted to get this online
so I just dumped my raw archive directory onto google drive... sorry!
Some assembly required!

I had a feeling that content might become unavailable due to the
increasingly buggy ill-maintained website and various other clues, so
I diligently archived everything as soon as it was released all the
way to the end, particularly throughout ksh93v- development. The final
snapshots contain a number of noteworthy changes, some of which I
think ought to be backported if possible. E.g. (off the top of my
head) the bash-compatible `type / whence` implementation is rather
nice and there's even a seemingly feature-complete implementation of
`compgen`.

A few highlights -
https://gist.github.com/ormaaj/3d296c4dc865f2a58554c438ce27da32

I haven't had much chance to follow the last year or so of work so I'm
sure some things have changed. :)

**Comment by ormaaj** [comment]
http://gsf.cococlyde.org/download

Glenn Fowler's site appears to still be online, hosting some of those files. I had this bookmarked for ages. Not sure if anyone found this.

**Comment by JohnoKing** [comment]
> The final snapshots contain a number of noteworthy changes, some of which I
think ought to be backported if possible. E.g. (off the top of my
head) the bash-compatible `type / whence` implementation is rather
nice and there's even a seemingly feature-complete implementation of
`compgen`.

Regarding `whence/type`, I backported the `-t` and `-P` flags quite a while ago in a separate branch:
https://github.com/JohnoKing/ksh/commit/9c9fbc7b49b108cf2c4ae23e15927e4a7f725360
I didn't submit that patch in a pull request because at the time the 93u+m [README](https://github.com/ksh93/ksh/blob/a065558291a112989035b47a88cfccfc380d38bc/README.md) had this statement in the policy:
> No new features; bug fixes only (but see items 3 and 4). Feature development is for a future separate branch.

The README file since then has been updated to state feature additions are for 2.x and up, so I've been holding onto the patch for until then. That said, I am willing to submit the feature in a pull request if @McDutchie says he's okay with it.

As far as `compgen` and `complete` are concerned, I did attempt to backport those at around the same I backported `type -t`, but I gave up on it because those 93v- builtins were just way too buggy.

**Reply by ormaaj** [reply]
Ha. Right by "nice implementation" i meant "I ran it and it worked". Some of that compatibility stuff was pretty ugly IIRC, like the `(exec -a bash ksh)` mode with the shell code "templates".

I guess I don't really care about getting in features urgently. I think @McDutchie is just trying to get something stable released which is the reason for that "policy" note that I would assume is temporary. Maybe if people have feature work and backports those can be consolidated into new branches for eventual merging.

**Reply by McDutchie** [reply]
> Regarding `whence/type`, I backported the `-t` and `-P` flags quite a while ago in a separate branch:
> [JohnoKing@9c9fbc7](https://github.com/JohnoKing/ksh/commit/9c9fbc7b49b108cf2c4ae23e15927e4a7f725360)
> I didn't submit that patch in a pull request because at the time the 93u+m [README](https://github.com/ksh93/ksh/blob/a065558291a112989035b47a88cfccfc380d38bc/README.md) had this statement in the policy:
[…]
> That said, I am willing to submit the feature in a pull request if @McDutchie says he's okay with it.

I'm okay with it, please go ahead. I think some users might quite appreciate having `type -t` from bash. Also, you could see it as improving an existing feature, namely `whence`.

You might have noticed I weakened that policy item to say that *major* feature development is for 2.x and up. We've been incrementally adding functionality for a while now. In general, if it improves or completes an existing feature, makes sense, looks useful, and can be added cleanly and without refactoring a bunch of things, I'm fine with it. Such things can go into 1.0 as well.

And we've now got a development branch that is separate from the 1.0 branch, so it might be about time to let some more substantial new features in as well – those that are for 1.1 and up.

**Comment by JohnoKing** [comment]
Just to give a quick update, I have created an ast-open-archive repo with all of the versions of ast-open and ksh93 that I could find (including the versions in ast-open-history and the various versions of 93v-). Links to my archival work so far:

https://github.com/ksh93/ast-open-archive
https://github.com/ksh93/ksh93-history
https://archive.org/details/ast-open-archive

**Reply by McDutchie** [reply]
That's fantastic, I'm impressed. Do you want to host these under the ksh93 github organisation? The ast-open-history repo is now superfluous, yours should replace it.

**Reply by JohnoKing** [reply]
I've went ahead and transferred the git repos to be under the ksh93 organization. That said, I'll ask you **not** to use github's archive repo feature on either repository. This is because I'd like to add more versions to each repo, should a tarball for any missing version of ast-open and/or ksh be found in the future (I used a script to generate both repos, so it isn't too difficult for me to update the git history).

**Reply by McDutchie** [reply]
OK, that's fine, though I'll note there's a fairly prominent "Unarchive this repository" button in the settings of the ast-open-history repo – which I'll now go and delete as it's redundant.

**Comment by ormaaj** [comment]
@JohnoKing Did you know about this? https://github.com/weiss/original-bsd/tree/master/local/toolchest/ksh It's apparently the ancient ksh88 history. I had no idea this was available. @izabera unearthed this and brought it to my attention. No clue how she found it, but very interesting!

**Reply by izabera** [reply]
https://minnie.tuhs.org/pipermail/tuhs/2020-December/022640.html
